### ISSUE TITLE
How to make Result to force me to not use result.Value if result.IsFailed==true

### ISSUE THREAD

**Issue by tesar-tech** [jscarle/LightResults#17]
Consider this:

```csharp
Resutl<MyClass> result = _service.GetMyClass();//is set to failed path
Console.WriteLine($"{result.Value.SomeProperty}");
```

Which will fail with `Result is failed. Value is not set.`, which is logical, right. 

My concern is that I shouldn't be allowed to shoot myself in my leg or IDE should be able to tell. 

Similarly like nullable is handled in .net.

**Comment by Joe4evr** [comment]
If you own this `Result` type yourself, you can solve it at the API-level by not exposing `Value` and instead having
```cs
public bool IsSuccess([MaybeNullWhen(false)] out T value);
```
And let nullable analysis do the work for you.

**Comment by jscarle** [comment]
> This could be solved at the API-level if instead of exposing `Value` directly it was
> 
> ```cs
> public bool IsSuccess([MaybeNullWhen(false)] out T value);
> ```
> 
> And let nullable analysis do the work for you.

I really like this approach. Thanks for the suggestion @Joe4evr!

**Comment by tesar-tech** [comment]
I got deeper into this issue and unfortunately cannot find any discussion on this topic. 

`FluentResult` have it similar - no guard for accessing the `Value`, same goes for `Ardalis.Result` and `ErrorOr` libs. 

The only one that "takes care" of this issue is `LanguageExt`, which doesn't expose the `Value`, you can only access it through `Match` or similar methods. 

My previous "result pattern" was `(MyClass?, ActionResult)`  tuple and flow similar to this:

```csharp
var (myclass, res) = _service.GetMyClass();//is set to failed path
if(myclass is null) return res;
Console.WriteLine($"{result.Value.SomeProperty}");
```
which will give me warning if I omit the `if`. 

Nick Chapsas [has thi video](https://www.youtube.com/watch?v=YbuSuSpzee4) about Result pattern, where his own implementation also exposes the `Value` only in the `Match` method. 

That's the functional way I guess, but it doesn't allow me an early return, thus doesn't allow me to  keep happy path without brackets and parenthese. 

I am sorry for having bit off-topic discussion. I am new to LightResults same as Result pattern in general.

**Comment by jscarle** [comment]
@tesar-tech I'm currently working on implementing what was suggested to do what you want. Just need to iron out the details and I'll publish the changes today in the preview version.

**Comment by jscarle** [comment]
The changes have been published as https://www.nuget.org/packages/LightResults/9.0.0-preview.3

**Comment by tesar-tech** [comment]
> The changes have been published as https://www.nuget.org/packages/LightResults/9.0.0-preview.3

Nice. I appreciate the pace!
I have it installed and it works amazing. It is even better than I expected, because now the flow can continue with the `MyClass` directly (no need for the Value at all).

Few consideration:

-  This is obviously breaking change that will make developer to re-do their usage of `Value` property
-  Now it has created kind of inconsistence between `IsFail` and `IsSuccess`.

My proposal: 
- Keep both `IsFail` and `IsSucess` as properties (as is was before)
- Introduce either: 
  - `Succeeded` and `Failed` methods, or:
  - `TryGetValue`, which keeps consistency with other types like `int.TryGetValue`, but doesn't allow to have the early return on failed state without negation `if(!TryGetValue` vs `if(Failed`. Also TryGetFailedValue is possible, but that's even more confusing.

I am able to help!

**Comment by MrDave1999** [comment]
@tesar-tech I don't understand. Who would do such a thing? The consumer should handle the error in this way:
```cs
Result<MyClass> result = _service.GetMyClass(); // is set to failed path
if(result.IsSuccess)
{
    Console.WriteLine($"{result.Value.SomeProperty}");
}
else 
{
    Console.WriteLine("Failed!");
}
```
The purpose of the result pattern is precisely to have an explicit and simple error handling. It is very strange that someone would directly access the Value property without explicit error handling.

**Comment by jscarle** [comment]
I just published [9.0.0-preview.4](https://www.nuget.org/packages/LightResults/9.0.0-preview.4) which brings several changes.

You can now do all of the following:

```cs
if (result.IsSuccess())
{
    // Do something
}
```

```cs
if (result.IsSuccess(out var value))
{
    // Do something with value
}
```

```cs
if (result.IsFailed())
{
    // Do something
}
```

```cs
if (result.IsFailed(out var error))
{
    // Do something with first error
}
```

I've also removed direct access to the `Value` and `Error` properties.

The `Result<TValue>.Ok()` and `Result<TValue>.Fail()` methods have also been removed for consistency, instead use the methods `Result.Ok<TValue>()` and `Result.Fail<TValue>()`.

**Comment by tesar-tech** [comment]
Yes. This makes much more sense to have the method overloads.

Also keeping the error only for the failed state is a good idea. 

The only problem I see is this:

```csharp
if (result.IsSuccess(out var value))
{
    // Do something with value
}
else
{
//don't have access to error, thus:
if(result.IsFail(out var error))
{
 //do something with error
}
}
```

This forces user to do the second (unnecessary) check, which is there only for getting the error. 

What about adding one more overload for IsSuccess and IsFail:

` ` ` csharp
 public bool IsSuccess([MaybeNullWhen(false)] out T value, [MaybeNullWhen(true)] out IError error)
 {
//...
 }
` ` ` 

would be usable as:

` ` ` csharp
if (result.IsSuccess(out var value, out var error))
{
    // Do something with value
}
else
{
//do something with error
}
` ` ` 

(and the "same" for IsFailed)

And everybody has what is needed. Just checking the result? `IsSuccess()` . Want also the result? `IsSuccess(out var value)`. Need to work with the error too? ` IsSuccess(out var value, out var error)`  

Note that this will not let you access value in unhappy path and the error in happy path => This is basically  what `Match` methods from other Result patter implementations do,  but (imo) in a better way, leveraging the already existing nullability feature in .net.

**Comment by jscarle** [comment]
I'm still reflecting on the best approach.

**Comment by jscarle** [comment]
Ok, so I did the following:

- Added additional overloads and methods for `Result`:
  - `void Match(Action success, Action<IError> failure)`
  - `TResult Match<TResult>(Func<TResult> success, Func<IError, TResult> failure)`
- Added additional overloads and methods for `Result<TValue>`:
  - `IsSuccess(out TValue value, out IError error)`
  - `IsFailed(out IError error, out TValue value)`
  - `void Match(Action<TValue> success, Action<IError> failure)`
  - `TResult Match<TResult>(Func<TValue, TResult> success, Func<IError, TResult> failure)`
- Added extension method:
  - `Result<TDestination> Bind<TSource, TDestination>(this Result<TSource> source, Func<TSource, Result<TDestination>> transform)`

You can now do things like:
```cs
Result<int> intResult = Result.Ok(0);
Result<string> bindResult = intResult.Bind(value => Result.Ok(value.ToString()));
bool matchResult = bindResult.Match(success => true, error => false);
```

It's now published as [9.0.0-preview.6](https://www.nuget.org/packages/LightResults/9.0.0-preview.6).
